src/vos/vos_gc.c

```diff
@@ -0,0 +1,832 @@
+	}
+
+	D_ASSERT(item->it_addr != 0);
+	rc = gc->gc_drain(gc, pool, item, &creds, empty);
+	if (rc)
+		return rc;
+
+	if (gc->gc_type == GC_AKEY) {
+		/* single value/recx tree wouldn't be flatterned (might be
+		 * changed in the future), instead they are freed within
+		 * dbtree/evtree_drain(), so user credits should be consumed.
+		 */
+		D_ASSERT(*credits >= creds);
+		*credits = creds;
+	}
+	return 0;
+}
+
+static int
+gc_free_item(struct vos_gc *gc, struct vos_pool *pool, struct vos_gc_item *item)
+{
+	struct vos_gc_bin_df *bin = gc_type2bin(pool, gc->gc_type);
+	struct vos_gc_bag_df *bag;
+	int		      first;
+	int		      rc = 0;
+
+	bag = umem_off2ptr(&pool->vp_umm, bin->bin_bag_first);
+	D_ASSERT(bag && bag->bag_item_nr > 0);
+	D_ASSERT(item == &bag->bag_items[bag->bag_item_first]);
+
+	first = bag->bag_item_first + 1;
+	if (first == bin->bin_bag_size)
+		first = 0;
+
+	if (first == bag->bag_item_last) {
+		/* it's going to be a empty bag */
+		D_ASSERT(bag->bag_item_nr == 1);
+		rc = gc_bin_free_bag(&pool->vp_umm, bin, bin->bin_bag_first);
+	} else {
+		rc = umem_tx_add_ptr(&pool->vp_umm, bag, sizeof(*bag));
+		if (rc)
+			goto failed;
+
+		bag->bag_item_first = first;
+		bag->bag_item_nr--;
+	}
+
+	D_DEBUG(DB_TRACE, "GC released a %s\n", gc->gc_name);
+	/* this is the real container|object|dkey|akey free */
+	if (gc->gc_free)
+		gc->gc_free(gc, pool, item);
+	else
+		umem_free(&pool->vp_umm, item->it_addr);
+
+	switch (gc->gc_type) {
+	default:
+		D_ASSERT(0);
+		break;
+	case GC_AKEY:
+		pool->vp_gc_stat.gs_akeys++;
+		break;
+	case GC_DKEY:
+		pool->vp_gc_stat.gs_dkeys++;
+		break;
+	case GC_OBJ:
+		pool->vp_gc_stat.gs_objs++;
+		break;
+	case GC_CONT:
+		pool->vp_gc_stat.gs_conts++;
+		break;
+	}
+failed:
+	return rc;
+}
+
+/**
+ * Add an item for garbage collection, this item and all its sub-items will
+ * be freed by vos_gc_run.
+ *
+ * NB: this function must be called within pmdk transactoin.
+ */
+int
+gc_add_item(struct vos_pool *pool, enum vos_gc_type type, umem_off_t item_off,
+	    uint64_t args)
+{
+	struct vos_gc_bin_df *bin = gc_type2bin(pool, type);
+	struct vos_gc_item    item;
+
+	D_DEBUG(DB_IO, "Add %s addr="DF_X64"\n", gc_type2name(type), item_off);
+	item.it_addr = item_off;
+	item.it_args = args;
+	while (1) {
+		int	  creds = GC_CREDS_PRIV;
+		int	  rc;
+		bool	  empty;
+
+		rc = gc_bin_add_item(&pool->vp_umm, bin, &item);
+		if (rc == 0) {
+			if (!gc_have_pool(pool))
+				gc_add_pool(pool);
+			return 0;
+		}
+
+		/* this is unlikely, but if we cannot even queue more items
+		 * for GC, it means we should reclaim space for this pool
+		 * immediately.
+		 */
+		if (rc != -DER_NOSPACE) {
+			D_CRIT("Failed to add item, pool="DF_UUID", rc=%s\n",
+			       DP_UUID(pool->vp_id), d_errstr(rc));
+			return rc;
+		}
+
+		if (d_list_empty(&pool->vp_gc_link)) {
+			D_CRIT("Pool="DF_UUID" is full but nothing for GC\n",
+			       DP_UUID(pool->vp_id));
+			return rc;
+		}
+
+		rc = gc_reclaim_pool(pool, &creds, &empty);
```

Comment:
Why we need to call a gc_reclaim_pool() when adding item? Won't GC ULT call the gc_reclaim_pool()?

Reply:
This can only happen when it's under free space pressure (cannot even add item, very unlikely), so we try to reclaim space and try again.

Reply:
I see, then I think we'd choose a larger credit here to make sure real free happened on value tree? (instead of just flatten cont/obj/dkey trees which won't actually reclaim much space)
Actually I guess a larger value could be used for default credit (current is 128) for efficiency.